src/components/ec/cuda/ec_cuda.h

```diff
@@ -142,8 +143,11 @@ extern ucc_ec_cuda_t ucc_ec_cuda;
         cudaError_t cuda_st = cudaSuccess;                                     \
         ucc_spin_lock(&ucc_ec_cuda.init_spinlock);                             \
         if (ucc_ec_cuda.stream == NULL) {                                      \
-            cuda_st = cudaStreamCreateWithFlags(&ucc_ec_cuda.stream,           \
+            cudaStream_t _stream;                                              \
+            cuda_st = cudaStreamCreateWithFlags(&_stream,                      \
                                                 cudaStreamNonBlocking);        \
+            ucs_atomic_cswap64((uint64_t*)(uintptr_t)&ucc_ec_cuda.stream,      \
+                               (uint64_t)NULL, (uint64_t)_stream);             \
```

Comment:
why is this needed? seems there is already lock above

Reply:
the lock is not enough, there is a race:
consider 2 threads. 1st thread grabs the lock above and goes to streamCreate. Inside that API the stream gets allocated and the ptr != NULL, yet the stream object is not entirely initialized (we should not use the stream ptr until the return from the cudaStreamCreate call). However, another thread reads non-NULL stream value and uses it, e.g. in ucc_mc_memcpy. I've reproduced that in that PR with modified MT ucc_test_mpi. 

I use local variable to init stream + atomic to write 8bytes of ptr to shared ucc_ec_cuda.stream atomically, so that other threads could not read incorrect ptr.

Reply:
After discussion, change: using variable stream_initialized on ucc_mc/ec_cuda instead of atomic in order to avoid assumption sizeof(cudaStream_t) == 8.